rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,184 @@
+#include "rewriter/ChainedSig.h"
+#include "ast/Helpers.h"
+#include "ast/treemap/treemap.h"
+#include "core/errors/rewriter.h"
+
+using namespace std;
+namespace sorbet::rewriter {
+
+struct ChainedSigWalk {
+    // The previousSigSend is used to allow us to add errors to `sig` with missing block wihtout accidentally adding it
+    // to a chained sig. E.g.:
+    // `sig` -> error no block
+    // `sig.final {}` -> no error. Need to prevent adding the error when we receive `sig` by itself
+    ast::Send *previousSigSend;
+
+    ast::ExpressionPtr preTransformSend(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto *send = ast::cast_tree<ast::Send>(tree);
+
+        // Return early if we identify things like `params` being invoked on `sig`
+        if (invalidChainedStatement(ctx, send)) {
+            return tree;
+        }
+
+        if (send->fun == core::Names::sig() && send->block == nullptr && send != previousSigSend) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // Return early unless it's one of the sends we are interested in
+        if (send->fun != core::Names::abstract() && send->fun != core::Names::final_() &&
+            send->fun != core::Names::override_() && send->fun != core::Names::overridable()) {
+            return tree;
+        }
+
+        ast::Send *sigSend = ast::cast_tree<ast::Send>(send->recv);
+
+        // Make sure the first receiver is sig. E.g.: `sig.abstract {}`, `sig.override.final {}`
+        if (sigSend == nullptr || !firstReceiverIsSig(sigSend)) {
+            return tree;
+        }
```

Comment:
(from comment below):

concretely, can we move this stanza up to be before invalidChainedStatement, and delete the while loop from invalidChainedStatement?

Reply:
Done, got rid of the loop.